src/events/upgrade.py

```diff
@@ -0,0 +1,136 @@
+#!/usr/bin/env python3
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Upgrades implementation."""
+import logging
+from typing import TYPE_CHECKING
+
+from charms.data_platform_libs.v0.upgrade import (
+    ClusterNotReadyError,
+    DataUpgrade,
+    DependencyModel,
+    KubernetesClientError,
+)
+from lightkube.core.client import Client
+from lightkube.core.exceptions import ApiError
+from lightkube.resources.apps_v1 import StatefulSet
+from ops.framework import EventBase
+from ops.model import ModelError
+from pydantic import BaseModel
+from tenacity import retry, stop_after_attempt, wait_random
+from typing_extensions import override
+
+from literals import CONTAINER
+
+if TYPE_CHECKING:
+    from charm import ZooKeeperCharm
+
+logger = logging.getLogger(__name__)
+
+
+class ZooKeeperDependencyModel(BaseModel):
+    """Model for ZooKeeper Operator dependencies."""
+
+    service: DependencyModel
+
+
+class ZKUpgradeEvents(DataUpgrade):
+    """Implementation of :class:`DataUpgrade` overrides for in-place upgrades."""
+
+    def __init__(self, charm: "ZooKeeperCharm", **kwargs):
+        super().__init__(charm, **kwargs)
+        self.charm = charm
+        self.substrate = self.charm.state.substrate
```

Comment:
Why is this needed? There seems to be no usage of this in the class

Reply:
Good point! I'll remove it before merging.

Reply:
I was wrong to do this. We need the substrate, else the `pre-upgrade-check` defaults to "vm", and attempts to run a not-implemented `build_upgrade_stack`.

Will fix.